### PR TITLE
Fixing the norm constraint formula

### PR DESCRIPTION
**Pull request by laurent-dinh**

### CODE REVIEW

**Comment by JesseLivezey** [comment]
Would it make sense to factor out some of the computation into a ```pylearn2.expr.updates.py``` file or something similar so that the same/similar code isn't rewritten everywhere? This might also make testing easier.

**Comment by lamblin** [comment]
I agree with @JesseLivezey, could you factor it out? Especially if we want to later change the 1e-7 magical constant.

**Comment by memimo** [comment]
Looks like his new commit is not triggering travis. @lamblin do you know if what could be the cause?

**Comment by lamblin** [comment]
It might have been an issue of Travis itself. I'll close and re-open.

**Comment by lamblin** [comment]
Putting just the `max` is not that helpful... I would have expected something like:
```python
def renormalize_factor(desired_norms, norms, eps=1e-7):
    """
    (Helpful docstring here)
    """
    return desired_norms / T.maximum(norms, eps)
```
That way, we have more control over `eps` as well.

**Comment by goodfeli** [comment]
I already fixed this, in a factored way, for maxout. Please use the same MaxL2FilterNorm ModelExtension here too.

**Comment by goodfeli** [comment]
Here's the ModelExtension: https://github.com/lisa-lab/pylearn2/blob/c4293f99fa2028cc486512b8989e9c5c5fb87132/pylearn2/model_extensions/norm_constraint.py

**Comment by goodfeli** [comment]
Here's how I use it for maxout: https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/models/maxout.py#L172

**Comment by laurent-dinh** [comment]
Thank you Ian.
Although I am well aware of this ModelExtension, I am currently trying to address [this issue](https://github.com/lisa-lab/pylearn2/issues/1126) in this PR. As this PR is turning into refactoring the norm constraint, it's good that you brought that into the topic. Should we close this PR and focus on modifying and using this ModelExtension for all concerned models ?

**Comment by goodfeli** [comment]
Well, 1126 does say to do 1128 first.
Basically if you do 1128, then 1126 is no longer relevant because the ModelExtension is already fixed.

**Comment by lamblin** [comment]
@laurent-dinh OK, it would be better to use the existing ModelExtension, I forgot that it already existed.

File: pylearn2/models/mlp.py [hunk 1]
```diff
@@ -2024,15 +2020,6 @@ def _modify_updates(self, updates):
...
-        if self.max_row_norm is not None:
-            W, = self.transformer.get_params()
-            if W in updates:
-                updated_W = updates[W]
-                row_norms = T.sqrt(T.sum(T.sqr(updated_W), axis=1))
-                desired_norms = T.clip(row_norms, 0, self.max_row_norm)
-                scales = desired_norms / (1e-7 + row_norms)
-                updates[W] = updated_W * scales.dimshuffle(0, 'x')
-
         if self.max_col_norm is not None or self.min_col_norm is not None:
```
Comment (laurent-dinh) [inline pylearn2/models/mlp.py]: Should I integrate a minimum norm constraint ?
If so, should I forbid using row and column norm constraint at the same time.

Comment (memimo) [inline pylearn2/models/mlp.py]: yeah, let's integrate the minimum norm constraint.
And yes, you should forbid both row and column norm constraints at same time, that's what current code is doing now.

**Comment by goodfeli** [comment]
The minimum norm issue is separate from the row norm issue.

As far as I know, no one uses either max_row_norm or min_col_norm. Ideally, we'd mark them as deprecated for 6 months before removing them. In this case it might be OK to just remove them.

I wrote them while experimenting with what kind of norm constraints work well, and after deciding that max_col_norm was the only one that was consistently helpful I have quit using the other two. I don't know of anyone else using them.

File: pylearn2/model_extensions/norm_constraint.py [hunk 2]
```diff
@@ -21,12 +21,19 @@ class MaxL2FilterNorm(ModelExtension):
     Parameters
     ----------
     limit : float or symbolic float
-        The maximum column norm of the weight matrix is constrained
-        to be <= limit
+        The maximum norm of the weight matrix is constrained
+        to be <= limit along the axes
+    limit : float or symbolic float
```
Comment (memimo) [inline pylearn2/model_extensions/norm_constraint.py]: min_limit*

File: pylearn2/model_extensions/norm_constraint.py [hunk 3]
```diff
@@ -21,12 +21,19 @@ class MaxL2FilterNorm(ModelExtension):
...
+        The maximum norm of the weight matrix is constrained
+        to be => limit along the axes
```
Comment (memimo) [inline pylearn2/model_extensions/norm_constraint.py]: minimum

**Comment by memimo** [comment]
Except the minor doc comments LGTM.

**Comment by laurent-dinh** [comment]
Alright, the Travis error caused by an external library should be fixed now. 
Can someone relaunch the Travis test ?

**Comment by laurent-dinh** [comment]
Thanks !

**Comment by memimo** [comment]
@lamblin if you have no further comment, I will merge today.

File: pylearn2/model_extensions/norm_constraint.py [hunk 4]
```diff
@@ -21,12 +21,19 @@ class MaxL2FilterNorm(ModelExtension):
...
+    axis : int or tuple of int
+        The axis or axes over which the norm is computed. Default is 0.
     """
 
-    def __init__(self, limit):
-        self.limit = limit
+    def __init__(self, limit, min_limit=0., axis=0):
+        self.max_limit = limit
+        self.min_limit = min_limit
+        self.axis = axis
```
Comment (lamblin) [inline pylearn2/model_extensions/norm_constraint.py]: If both `min_limit` and `max_limit` are provided, can you assert that `min_limit <= max_limit`? The behaviour in the other case is ill-defined.

**Comment by lamblin** [comment]
Other than that, it works for me.

**Comment by laurent-dinh** [comment]
The modifications asked are done and the build passed.

**Pull request merged**